bootstrap/ssh.go

```diff
@@ -0,0 +1,54 @@
+package bootstrap
+
+import (
+	"os"
+	"path/filepath"
+	"runtime"
+
+	"github.com/buildkite/agent/bootstrap/shell"
+	"github.com/pkg/errors"
+)
+
+func sshKeygen(sh *shell.Shell, knownHostsfile, host string) (string, error) {
+	sshKeygen, err := sh.AbsolutePath("ssh-keygen")
+	if err != nil && runtime.GOOS == "windows" {
+		var winErr error
+		if sshKeygen, winErr = findPathToSSHToolsOnWindows(sh); winErr != nil {
+			return "", err
+		}
+	}
+
+	return sh.RunAndCapture(sshKeygen, "-f", knownHostsfile, "-F", host)
+}
+
+func sshKeyScan(sh *shell.Shell, host string) (string, error) {
+	sshKeyscan, err := sh.AbsolutePath("ssh-keyscan")
+	if err != nil && runtime.GOOS == "windows" {
+		var winErr error
+		if sshKeyscan, winErr = findPathToSSHToolsOnWindows(sh); winErr != nil {
+			return "", err
+		}
+	}
+
+	return sh.RunAndCapture(sshKeyscan, host)
```

Comment:
We need to capture stdout and stderr separately — only stdout should be appended to the known hosts file.

Later versions of ssh-keyscan will exit non-zero, but earlier versions just won't output anything to stdout. We should probably catch these cases as errors, or at least warnings.

Reply:
Hmmm. I'd say we should ignore stderr entirely in RunAndCapture then.

Reply:
Or always capture stdout and leave stderr connected to the actual stderr if Debug? Or something?